### PR TITLE
Static model preferences for Gateways

### PR DESCRIPTION
**Pull request by jhawthorn**

### CODE REVIEW

File: core/lib/generators/spree/install/templates/config/initializers/spree.rb [hunk 1]
```diff
@@ -36,6 +36,18 @@
 
   # Custom logo for the admin
   # config.admin_interface_logo = "logo/solidus_logo.png"
+
+  # Gateway credentials can be configured statically here and referenced from
+  # the admin. They can also be fully configured from the admin.
+  #
+  # config.static_model_preferences.add(
+  #   Spree::Gateway::StripeGateway,
+  #   'example_credentials',
```
Comment (adammathys) [inline core/lib/generators/spree/install/templates/config/initializers/spree.rb]: This is a little ambiguous. When I first read it, I thought it meant to put my credentials here. Not that this is a nice name for referencing these credentials later. Maybe something like `stripe_env_credentials` would be clearer.

**Comment by adammathys** [comment]
Also need to update [this](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/gateway.rb#L28-L30) to something like:

```ruby
def options
  preferences.to_hash
end
```

Because what's currently there is silly, it just converts all keys to symbols, and unnecessary. Also, it doesn't play nice with the static preference class.

**Comment by jhawthorn** [comment]
@adammathys thanks. Updated with your feedback.

**Comment by adammathys** [comment]
:+1: Awesome. All looks good to me.

**Comment by magnusvk** [comment]
This looks good to me, though I'd love another set of eyes from @jordan-brough or @athal7 as to how this would play with the custom work we've done around this ourselves.

Also, do we want documentation for this somewhere?

**Comment by adammathys** [comment]
@magnusvk, @jordan-brough, @athal7: For reference, [here's how this hooks into spree-backend](https://github.com/bonobos/spree-backend/blob/6943ec36cff6863d6186732942de4779cabfe8da/lib/bonobos/gateway_configuration.rb) right now. Which provides an almost identical interface to the other static preferences in `spree-backend`:

```ruby
Bonobos::GatewayConfiguration.new(
  SpreeBackend::Application::GATEWAYS_FILE
).load!
```

I do have a minor tweak or two to make to the load class. (Specifically, it won't be updating records since that can cause all sorts of issues.)

**Comment by magnusvk** [comment]
That looks great, thanks for sharing that.

**Comment by athal7** [comment]
i may be seeing this wrong, but it seems to me it's still possible to view and update preferences even when loading from a yml file? i think this is a good first step, but i'd probably want to not have the configs viewable in the UI or stored in the DB if they are specified via in-memory preferences

**Comment by jhawthorn** [comment]
> i may be seeing this wrong, but it seems to me it's still possible to view and update preferences even when loading from a yml file? i think this is a good first step, but i'd probably want to not have the configs viewable in the UI or stored in the DB if they are specified via in-memory preferences

For PaymentMethod which has a `preference_source` set (a string which identifies the static credentials to use) the settings should not be viewable and can't be edited.

**Comment by adammathys** [comment]
As with [spree_config_preferences](https://github.com/bonobos/spree_config_preferences), it's still possible to view the preferences in the admin when loading from a YAML file. However, it now hides all of the preference fields. Making it impossible to update them without first consciously deciding to convert to the database store preferences. (Also, the assignment operation is a no-op when using the static preferences.)
static

For reference, this is what it looks like when editing a gateway with preferences stored in the database:
dynamic

**Comment by athal7** [comment]
:+1:

**Pull request merged**